[MPS] Implemented `masked_fill_scalar` as shader

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147369

- Move `pos_from_thread_index and `offset_from_pos` from `UnfoldBackward.metal` into `c10/metal/indexing.h` header
- Initial idea were to implement `StridedTensor` and `ConstStridedTensor` and use them to have masked_fill kernel a something simple as the following loop
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```
But though it looks elegant and works correctly, performance wise it's much slower that the existing MPS shader (see table below), as int64 divisions on M2 GPU are really slow

- Solved performance issue by implementing 3 flavors of the same shader: `dense`, that is used when both input and mask are dense tensors of the same size, `broadcast`, which is used when `mask` is leading dimensions expandable into input tensor and `strided`  which is a general purpose fallback, but still computes position in the tensors only ones. As result, perf is even better than existing MPS shader for dense and broadcast able tensors.

Performance measured on M2Pro thru different iterations of the same shader

| dtype | MPS | int64-idx | int64-inlined | 32-bit strided | 32-bit broadcasted |
| ------|------| -----|   ---- | --- | ---- |
| float32 | 2.8 msec  | 41.6 msec | 26.9 msec | 5 msec | 2.4 msec |
| float16 | 1.86 msec | 38.2 msec| 26.6 msec | 4.6 msec | 1.9 msec |
|bfloat16|1.86 msec |38.3 msec | 26.6 msec | 4.6 msec | 1.9 msec |

And benchmark script
```python
import torch

from timeit import default_timer
from itertools import product
from torch.utils.benchmark import Measurement, Timer

def bench_mask_fill(
    n,
    binary_func,
    dtype=torch.float32,
) -> Measurement:
    t = Timer(
        stmt=f"x.masked_fill(y, -17.0); torch.mps.synchronize()",
        setup=f"x,y = torch.rand(1, 20, {n}, {n}, dtype={dtype}, device='mps'), torch.ones({n}, {n}, device='mps').triu().bool()",
        globals = {'f': binary_func},
        language="python", timer=default_timer
    )
    return t.blocked_autorange()

if __name__ == "__main__":
    n = 1024
    for dtype in [torch.float32, torch.float16, torch.bfloat16]:
        eager_t = bench_mask_fill(n, torch.fmax, dtype)
        use_msec = eager_t.mean > 1e-4
        multiplier = 1e3 if use_msec else 1e6
        uname = "msec" if use_msec else "usec"
        print(f"torch.masked_fill_() {str(dtype):>14} {eager_t.mean*multiplier:>7.2f} {uname}")
```
Fixes https://github.com/pytorch/pytorch/issues/143477

malfet committed:
[MPS] Implemented masked_fill_scalar as shader

Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement StridedTensor and ConstStridedTensor

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/native/mps/kernels/UnfoldBackward.metal:
```diff
@@ -52,7 +32,7 @@ kernel void unfold_backward(
   auto step = dim_size_step_ndim.z;
   auto ndim = dim_size_step_ndim.w;
-  long pos[16];
+  long pos[max_ndim];
```
Doesn't metal support templated arrays? Any reason we are not using them here? Thought it was something like:
```suggestion
  array<long,max_ndim> pos;
```

malfet replied:
Alas, not it does not seem to work
```
In file included from /Users/malfet/git/pytorch/pytorch/aten/src/ATen/native/mps/kernels/Indexing.metal:1:
/Users/malfet/git/pytorch/pytorch/c10/metal/indexing.h:39:3: error: no matching function for call to 'pos_from_thread_index'
  pos_from_thread_index(idx, pos, sizes, ndim);
  ^~~~~~~~~~~~~~~~~~~~~
/Users/malfet/git/pytorch/pytorch/c10/metal/indexing.h:22:13: note: candidate function not viable: no known conversion from '::metal::array<long, max_ndim>' to 'long *' for 2nd argument
```

Skylion007 replied:
@malfet I see. You might need to unbox it with `pos.data()` or something, but yeah.

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet commented:
But perf is not great according to
```python
import torch

from timeit import default_timer
from itertools import product
from torch.utils.benchmark import Measurement, Timer

def bench_mask_fill(
    n,
    binary_func,
    dtype=torch.float32,
) -> Measurement:
    t = Timer(
        stmt=f"x.masked_fill(y, -17.0); torch.mps.synchronize()",
        setup=f"x,y = torch.rand(1, 20, {n}, {n}, dtype={dtype}, device='mps'), torch.ones({n}, {n}, device='mps').triu().bool()",
        globals = {'f': binary_func},
        language="python", timer=default_timer
    )
    return t.blocked_autorange()

if __name__ == "__main__":
    n = 1024
    for dtype in [torch.float32, torch.float16, torch.bfloat16]:
        eager_t = bench_mask_fill(n, torch.fmax, dtype)
        use_msec = eager_t.mean > 1e-4
        multiplier = 1e3 if use_msec else 1e6
        uname = "msec" if use_msec else "usec"
        print(f"torch.masked_fill_()x3 {str(dtype):>14} {eager_t.mean*multiplier:>7.2f} {uname}")
```

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Implemented masked_fill_scalar as shader"


Move existing pos_from_thread_index and offset_from_pos from
UnfoldBackward into `c10/metal/indexing.h` header

Implement `StridedTensor` and `ConstStridedTensor`, which could be used to implement masked_fill as something simple as
```metal
ConstStridedTensor<bool> mask(mask_data, sizes, mask_strides, ndim);
if (mask[thread_index]) {
  StridedTensor<T> input(input_data, sizes, input_strides, ndim);
  input[thread_index] = val;
}
```

Fixes https://github.com/pytorch/pytorch/issues/143477

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Lint + MPS are green"
